consensus/src/chained_bft/event_processor.rs

```diff
@@ -501,26 +505,60 @@ impl<T: Payload> EventProcessor<T> {
     /// 3. In case a validator chooses to vote, send the vote to the representatives at the next
     /// position.
     async fn process_proposed_block(&mut self, proposal: Block<T>) {
+        // Safety invariant: For any valid proposed block, its parent block == the block pointed to
+        // by its QC.
+        debug_checked_precondition_eq!(
+            proposal.parent_id(),
+            proposal.quorum_cert().certified_block_id()
+        );
+        // Safety invariant: QC of the parent block is present in the block store
+        // (Ensured by the call to pre-process proposal before this function is called).
+        debug_checked_precondition!(self
+            .block_store
+            .get_quorum_cert_for_block(proposal.parent_id())
+            .is_some());
+
         if let Some(time_to_receival) =
             duration_since_epoch().checked_sub(Duration::from_micros(proposal.timestamp_usecs()))
         {
             counters::CREATION_TO_RECEIVAL_S.observe_duration(time_to_receival);
         }
 
         let proposal_round = proposal.round();
+        // Creating proposal_id here since proposal gets moved in the call to execute_and_vote.
+        // Used in MIRAI annotation later.
+        let proposal_id = proposal.id();
+        let proposal_parent_id = proposal.parent_id();
+
         let vote_msg = match self.execute_and_vote(proposal).await {
             Err(_) => {
                 return;
             }
             Ok(vote_msg) => vote_msg,
         };
 
+        // Safety invariant: The vote being sent is for the proposal that was received.
+        debug_checked_verify_eq!(proposal_id, vote_msg.block_id());
+        // Safety invariant: The last voted round is updated to be the same as the proposed block's
+        // round. At this point, the replica has decided to vote for the proposed block.
+        debug_checked_verify_eq!(
+            self.safety_rules.consensus_state().last_vote_round(),
+            proposal_round
+        );
+
```

Comment:
two questions: 

* shouldn't all of these actually panic if not respected? (so be `checked_verify`)
* shouldn't all of these happen before we actually generate the vote? (`execute_and_vote()` above)

Reply:
> * shouldn't all of these actually panic if not respected? (so be `checked_verify`)

`debug_checked_verify` causes these to panic when the tests are run in the debug mode. We think this might not be the right time to add these checks in the release mode (considering performance) but we can discuss further. Eventually the goal is for MIRAI to be able to verify these checks statically.   

> * shouldn't all of these happen before we actually generate the vote? (`execute_and_vote()` above)

This invariant is like a postcondition for `execute_and_vote()`. `vote_msg` is generated by this function and therefore these invariants must be satisfied after its execution.

Reply:
I'm torn on this, on one hand there should still be some runtime insurance that we're not going to vote on something wrong; on the other hand is crashing really the correct solution when something like this happens, what about just returning. I'm wondering what consensus thinks about this, @dmitri-perelman @aching ?

Reply:
We probably should panic when such condition violated, it's a bug that needs fix. Returning would just hide the problem.

Reply:
Although I'm not sure if we could leave enough information for debug

Reply:
Indeed, but that is beyond the scope of this PR.

Reply:
@zekun000 what about having an alert that we can catch if this bug happens, and returning? Keep in mind that crashing here would mean crashing the whole network. No more transactions happening for all of our users. In both solutions we learn that there is a bug, in only one of them we can perhaps still function.

Reply:
@hermanventer yeah sorry for hijacking the PR, this can definitely wait for another PR.